### PR TITLE
Simplify choices for building LAMMPS

### LINKED ISSUE #1977: [Feature Request] Simplify build options by always building the LAMMPS library
**Summary**

Right now we have essentially 4 main build options:
- build executable from "static" objects (= no -fpic)
- build executable from "shared" objects (= with -fpic)
- build a static library
- build a shared library

I propose to reduce this to two options: always build the executable and the library, and only select between a shared library and a static library.

For CMake that means to get rid of the `BUILD_EXE` and `BUILD_LIB` options and only switch between `BUILD_SHARED_LIBS` set to **on** or **off**

For the traditional make, this would correspond to replace "mode=exe", "mode=shexe", "mode=lib", "mode=shlib" with "mode=static" and "mode=shared" (and for backward compatibility map "exe" and "lib" to "static" and "shexe" and "shlib" to "shared"). The default would be "static" so the result would be essentially the same except for creating the library in addition.

P.S.: anybody know why `BUILD_SHARED_LIBS` is plural? we only build one library and we enable it with `BUILD_LIB` (not `BUILD_LIBS`), which feels inconsistent.

### PR DESCRIPTION
**Pull request by akohlmey**
**Summary**

Change conventional and CMake build system to always build executable and library.
The only remaining choice is between shared library and static library.

**Related Issues**

This implements and thus closes #1977 

**Author(s)**

Axel Kohlmeyer (Temple U)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

For the conventional build the `mode=exe` and `mode=lib` options are forwarded to `mode=static` and `mode=shexe` and `mode=shlib` are forwarded to `mode=shared`.
For CMake setting the removed variables `BUILD_EXE` and `BUILD_LIB` will be ignored and there will be a warning to that effect at the end of the configuration step.
Any custom personal makefiles (e.g. in `src/MAKE/MINE`) need to be adapted. All bundled files have been adapted, but not all of them tested due to lack of access to suitable hardware.

**Implementation Notes**

TODO list:

- [x] modify `src/Makefile`
- [x] modify makefiles in `src/MAKE`
- [x] modify makefiles in `src/MACHINES`
- [x] modify makefiles in `src/OPTIONS`
- [x] update CMake related files
- [x] update `cmake/README.md`
- [x] update documentation
- [x] discuss that debug info (-g) is included by default and how to remove it (use strip or drop flag)
- [x] provide example/tutorial in Error docs for how to obtain a stack trace and use valgrind to identify a memory bug
- [x] explain C++11 compiler error message in Error docs

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] A package specific README file has been included or updated

### CODE REVIEW

**Comment by akohlmey** [comment]
@junghans please do not approve a pull request that is still a draft and listed as work in progress. thanks.

File: cmake/Modules/Packages/KIM.cmake [hunk 1]
```diff
@@ -51,6 +51,9 @@ if(DOWNLOAD_KIM)
     INTERFACE_INCLUDE_DIRECTORIES "${INSTALL_DIR}/include/kim-api")
   target_link_libraries(lammps PRIVATE LAMMPS::KIM)
   add_dependencies(LAMMPS::KIM kim_build)
+  if(BUILD_LIB AND NOT BUILD_SHARED_LIBS)
```
Comment (junghans) [inline cmake/Modules/Packages/KIM.cmake:54]: This `BUILD_LIB` needs to go.

File: cmake/Modules/Packages/KOKKOS.cmake [hunk 2]
```diff
@@ -22,6 +22,9 @@ if(DOWNLOAD_KOKKOS)
     INTERFACE_LINK_LIBRARIES ${CMAKE_DL_LIBS})
   target_link_libraries(lammps PRIVATE LAMMPS::KOKKOS)
   add_dependencies(LAMMPS::KOKKOS kokkos_build)
+  if(BUILD_LIB AND NOT BUILD_SHARED_LIBS)
```
Comment (junghans) [inline cmake/Modules/Packages/KOKKOS.cmake:25]: And here

File: cmake/Modules/Packages/LATTE.cmake [hunk 3]
```diff
@@ -27,6 +27,9 @@ if(DOWNLOAD_LATTE)
     INTERFACE_LINK_LIBRARIES "${LAPACK_LIBRARIES}")
   target_link_libraries(lammps PRIVATE LAMMPS::LATTE)
   add_dependencies(LAMMPS::LATTE latte_build)
+  if(BUILD_LIB AND NOT BUILD_SHARED_LIBS)
```
Comment (junghans) [inline cmake/Modules/Packages/LATTE.cmake:30]: And here.

File: cmake/Modules/Packages/KIM.cmake (hunk 1 above)
Comment (akohlmey) [reply cmake/Modules/Packages/KIM.cmake]: hmmm... that statement must have been added through merging your cmake branch after I had checked the cmake files. should be corrected now.

**Comment by sjplimp** [comment]
@akohlmey Please just add the doc page info we talked about: explain why size of exe is not what it seems, also how to strip it if desired, or remove -g switches (in make and cmake) if desired

**Comment by akohlmey** [comment]
> @akohlmey Please just add the doc page info we talked about: explain why size of exe is not what it seems, also how to strip it if desired, or remove -g switches (in make and cmake) if desired

@sjplimp did most of that over the weekend. Check out:
https://github.com/akohlmey/lammps/blob/simplify_build_choices/doc/src/Build_basics.rst

But I wanted to do a little more and added a section about getting stack traces from gdb and valgrind to the Errors section. https://github.com/akohlmey/lammps/blob/simplify_build_choices/doc/src/Errors_debug.rst
and also integrate relevant parts from @rbberger 's CMake tutorial and a few additional aspects of that. https://github.com/akohlmey/lammps/blob/simplify_build_choices/doc/src/Howto_cmake.rst

will do another pass at reviewing it and then merge unless there are objections.

**Comment by athomps** [comment]
I don't know if there is any way to avoid this, but people who pull from master and do 'make my_mpi' will find that everything compiles fine, but the linker fails with a message like:
Undefined symbols for architecture x86_64:
  "_main", referenced from:
It took me a while to realize I needed to update my low-level Makefile.

**Comment by akohlmey** [comment]
> I don't know if there is any way to avoid this, but people who pull from master and do 'make my_mpi' will find that everything compiles fine, but the linker fails with a message like:

if it could have been avoided, I would have done so. this is why I wanted the change now and not when we are close to the next stable release. there are several more changes with possible breakage in this patch (and going to be in the next). for example, if you do "make mode=shared mpi" you now also must set LD_LIBRARY_PATH. 

> Undefined symbols for architecture x86_64:
> "_main", referenced from:
> It took me a while to realize I needed to update my low-level Makefile.

you can see it as a not-so-subtle reminder to try out building LAMMPS with cmake, and thus part of the master plan for cmake's world domination. ;-)

**Pull request merged**